cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,145 @@
+
+namespace fs {
+
+// A system clock time point expressed as a 64-bit (or more) number of
+// secondes since the epoch.
+using TimePoint = std::chrono::sys_time<std::chrono::nanoseconds>;
+
+enum class FileType {
+  // Target does not exist
+  NonExistent,
+  // Target exists but its type is unknown (could be a special file such
+  // as a Unix socket or character device, or Windows NUL / CON / ...)
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+std::string ToString(FileType);
+
+struct FileStats {
+  FileType type = FileType::Unknown;
+  // The file path in the filesystem.
+  std::string path;
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size = -1;
+  // The time of last modification, if available.
+  TimePoint mtime = TimePoint(-1);
+
+  FileStats() = default;
+
+  // The file base name (component after the last '/').
+  std::string base_name() const;
+};
+
+struct OpenFlags {
+  // Whether to transparently compress or decompress, based on file extension
+  // or other heuristics.
+  bool transparent_compression = true;
+
+  OpenFlags() = default;
+};
+
+struct Selector {
+  // The folder in which to select files.
+  std::string base_folder;
+  // Whether to recurse into subfolders.
+  bool recursive = false;
+
+  MatchOptions() = default;
+};
+
+
+class FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively
+  virtual Status GetTargetStats(const std::string& path, FileStats* out);
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out) = 0;
+  virtual Status GetTargetStats(const Selector& select,
+                                std::vector<FileStats>* out) = 0;
+
+  /// Create a folder and subfolders.
+  virtual Status CreateFolder(const std::string& path, bool recursive = true) = 0;
+
+  /// Delete a folder and contents.
+  virtual Status DeleteFolder(const std::string& path) = 0;
+
+  /// Delete a file.
+  virtual Status DeleteFile(const std::string& path) = 0;
+  /// Delete many files.
+  virtual Status DeleteFiles(const std::vector<string>& path) = 0;
+
+  /// Move / rename a file or directory.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Move(const std::string& src, const std::string& dest) = 0;
+
+  /// Copy a file.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Copy(const std::string& src, const std::string& dest) = 0;
+
+  /// Open an input stream.
+  virtual Status OpenInputStream(const std::string& path,
+                                 std::shared_ptr<InputStream>* out);
+
+  virtual Status OpenInputStream(const std::string& path, OpenFlags flags,
+                                 std::shared_ptr<InputStream>* out) = 0;
```

Comment:
nit: should this be unique_ptr?

Reply:
I don't think so, `shared_ptr` is more flexible and we probably don't care about the performance benefit here.

Reply:
Yes, we pretty consistently use `shared_ptr` in these kinds of APIs